### DISCUSSION TITLE
CoreDNS performance

### DISCUSSION THREAD

**Discussion by dwbxm** [coredns/coredns#4625] [Q&A]
The performance of CoreDNS with loadbalance plugin degrades quickly.
The Corefile as follow:
```
. {
    hosts {
       10.1.1.2 test1.com
       fallthrough
    }
    etcd {
        path /net-services
        endpoint localhost:2379
        fallthrough
    }

    prometheus localhost:9153
    loadbalance 
    #unbound
    cache 300
    #reload 6s
    log
    errors
}
```

The result
```
[root@debug local-perf]# dnsperf -s 10.121.212.51 -l 10 -d test.txt 
DNS Performance Testing Tool
Version 2.3.2

[Status] Command line: dnsperf -s 10.121.212.51 -l 10 -d test.txt
[Status] Sending queries (to 10.121.212.51)
[Status] Started at: Thu May 13 08:55:15 2021
[Status] Stopping after 10.000000 seconds
[Status] Testing complete (time limit)


Statistics:

  Queries sent:         50424
  Queries completed:    50424 (100.00%)
  Queries lost:         0 (0.00%)

  Response codes:       NOERROR 50424 (100.00%)
  Average packet size:  request 27, response 62
  Run time (s):         10.002639
  Queries per second:   5041.069662

  Average Latency (s):  0.019695 (min 0.000328, max 0.569438)
  Latency StdDev (s):   0.085963
```

Now the Corefile changes:
```
. {
    hosts {
       10.1.1.2 test1.com
       fallthrough
    }
    etcd {
        path /net-services
        endpoint localhost:2379
        fallthrough
    }

    prometheus localhost:9153
    #4209 theloadbalance 
    #unbound
    cache 300
    #reload 6s
    log
    errors
}
```

The result is:
```
dnsperf -s 10.121.212.51 -l 10 -d test.txt 
DNS Performance Testing Tool
Version 2.3.2

[Status] Command line: dnsperf -s 10.121.212.51 -l 10 -d test.txt
[Status] Sending queries (to 10.121.212.51)
[Status] Started at: Thu May 13 09:12:24 2021
[Status] Stopping after 10.000000 seconds
[Status] Testing complete (time limit)

Statistics:

  Queries sent:         321770
  Queries completed:    321770 (100.00%)
  Queries lost:         0 (0.00%)

  Response codes:       NOERROR 321770 (100.00%)
  Average packet size:  request 27, response 62
  Run time (s):         10.001677
  Queries per second:   32171.604822

  Average Latency (s):  0.002861 (min 0.000295, max 0.010924)
  Latency StdDev (s):   0.000937
```

What result in this?

Environment:

coredns -version
CoreDNS-1.8.3
linux/amd64, go1.15.11,

lscpu
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                8
On-line CPU(s) list:   0-7
Thread(s) per core:    1
Core(s) per socket:    1
Socket(s):             8
NUMA node(s):          1

free -h
              total        used        free      shared  buff/cache   available
Mem:           7.6G        626M        2.5G        146M        4.5G        6.6G
Swap:            0B          0B          0B

Thanks.

**Comment by chrisohaver** [comment]
_loadbalance_ does add additional processing, but the difference you observed is much bigger than I would have guessed.
Per the average packet size, I presume that none of the answers are exceptionally long (long answers take more work to randomize).

Can you test with `resperf`?  https://linux.die.net/man/1/resperf .   The way `dnsperf` dynamically "self-paces" sometimes produces odd results - especially when testing for short intervals. CoreDNS processes requests in parallel, and this, i think, can sometimes confuse the self-pacing algorithm.

**Reply by dwbxm** [reply]
Thank you for your help.
Yes, the answers are not long. The CoreDNS server gives the result:
=====================================================
;; opcode: QUERY, status: NOERROR, id: 56290
;; flags: qr aa rd; QUERY: 1, ANSWER: 3, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;test1.com.	IN	 A

;; ANSWER SECTION:
test1.com.	3600	IN	A	5.5.5.9
test1.com.	3600	IN	A	5.5.5.8
test1.com.	3600	IN	A	10.1.1.2
=====================================================

I think that the difference between with loadbalance and without  loadbalance is not too big.
But the result is beyond imagination.
And dnsperf has the problem, so i will try resperf to test it. **what are he resperf'data file format ?** 

Dnsperf uses the test data "test.txt"'s contents are:
====================================================
cat test.txt 
test1.com A
test2.com A
test3.com A
test4.com A
test5.com A
====================================================

Thanks.

**Reply by chrisohaver** [reply]
> what are he resperf'data file format ?

I just noticed that you had asked this.  I don't know off the top of my head.  I don't use it frequently to have that memorized.

PS: You might want to read up on "markdown", which is what GitHub uses to format posts.
